### PR TITLE
Add a performance spec

### PR DESCRIPTION
**Pull request by mriddle**
There's no mention of the overhead that this gem introduces and I wanted a benchmark before adding the validation code in #63 

This PR also includes some minor refactoring, more detail can be found in the associated commits.

### CODE REVIEW

File: test/global_uid_test.rb [hunk 1]
```diff
@@ -205,6 +209,33 @@ def table_exists?(connection, table)
...
+        x.report(WithGlobalUID)    { WithGlobalUID.create! }
+        x.report(WithoutGlobalUID) { WithoutGlobalUID.create! }
+        x.compare!
+      end
+
+      with_global_uid    = report.entries.find { |e| e.label == WithGlobalUID }.stats
+      without_global_uid = report.entries.find { |e| e.label == WithoutGlobalUID }.stats
+
+      performance_impact, _ = with_global_uid.slowdown(without_global_uid)
+      assert performance_impact < 2
```
Comment (bquorning) [inline test/global_uid_test.rb:229]: Does this mean we would accept a factor 2 performance difference? I think we can go lower – maybe 1.5?

File: test/test_helper.rb [hunk 2]
```diff
@@ -9,14 +9,6 @@
...
 require 'pry'
 
-GlobalUid::Base.global_uid_options = {
-  :disabled   => false,
-  :id_servers => [
-    "test_id_server_1",
-    "test_id_server_2"
-  ]
-}
-
```
Comment (bquorning) [inline test/test_helper.rb:19]: Don’t we need to call `restore_defaults!` here?

File: test/global_uid_test.rb [hunk 3]
```diff
@@ -394,15 +394,14 @@ def parent_child_fork_values
...
-    (0..10).each do
-      foo = WithGlobalUID.new
-      foo.save
-      refute_nil foo.id
-      assert_nil foo.description
-      refute seen.has_key?(foo.id)
-      seen[foo.id] = 1
+  def test_unique_ids(amount = 10)
+    amount.times.each_with_object({}) do |_, seen|
+      WithGlobalUID.create!.tap do |record|
```
Comment (bquorning) [inline test/global_uid_test.rb:430]: Do we need the `tap` here? I would understand if this was inside a `map` block, but inside `each_with_object` the return value of the block doesn’t matter. So perhaps just `record = WithGlobalUID.create!` here?

File: test/models.rb [hunk 4]
```diff
@@ -3,6 +3,7 @@ class WithGlobalUID < ActiveRecord::Base
 end
 
 class WithoutGlobalUID < ActiveRecord::Base
+  disable_global_uid
```
Comment (bquorning) [inline test/models.rb:6]: I guess some tests have been testing the wrong thing until now?

File: test/global_uid_test.rb [hunk 5]
```diff
@@ -205,6 +205,33 @@ def table_exists?(connection, table)
...
+    end
+
+    it "has a negligible performance impact" do
+      report = Benchmark.ips do |x|
+        x.report(WithGlobalUID)    { WithGlobalUID.create! }
+        x.report(WithoutGlobalUID) { WithoutGlobalUID.create! }
+        x.compare!
+      end
+
+      with_global_uid    = report.entries.find { |e| e.label == WithGlobalUID }.stats
```
Comment (bquorning) [inline test/global_uid_test.rb:225]: This is private API, right? Do we risk that it breaks in the future. Consider adding a fixed version number in the gemspec.

File: test/models.rb (hunk 4 above)
Comment (mriddle) [reply test/models.rb:6]: Prior to this we were only testing the `MigrationExtension` with it disabled and the impact it had on the generated tables, not the model.

File: test/global_uid_test.rb (hunk 3 above)
Comment (mriddle) [reply test/global_uid_test.rb:430]: You're right, we don't need the tap. The method went under a couple of iterations and it was left behind 😅

File: test/global_uid_test.rb (hunk 1 above)
Comment (mriddle) [reply test/global_uid_test.rb:229]: There's a lot of variance between each run, I've seen it perform between 1.1x and 1.8x slower with it usually averaging around 1.3x. I put 2 here so the build doesn't break and anything higher than that is a cause for concern.

I'll see if I can ratchet the value down without adding noise to CI. That was my original hope when using benchmark-ips over benchmark

File: test/test_helper.rb (hunk 2 above)
Comment (mriddle) [reply test/test_helper.rb:19]: No, that's being called in the `before` block of the `GlobalUid` test

File: test/global_uid_test.rb (hunk 5 above)
Comment (mriddle) [reply test/global_uid_test.rb]: The [code is documented](https://github.com/evanphx/benchmark-ips/blob/0bb23ea1f5d8f65416629505889f6dfc550fa4a6/lib/benchmark/ips/report.rb#L132) so I took that as it being part of the public API. Although it doesn't seem like using reports directly is common, with that in mind, do you think it's still worth locking down the version?

File: test/global_uid_test.rb [hunk 6]
```diff
@@ -35,6 +35,17 @@
       assert !ParentSubclass.global_uid_disabled
       assert !ParentSubclassSubclass.global_uid_disabled
     end
+
+    it "uses the default AUTO_INCREMENT, skipping the alloc servers" do
+      CreateWithoutGlobalUIDs.up
+      GlobalUid::Base.expects(:get_uid_for_class).never
+
+      (1..10).each do |index, _|
```
Comment (bquorning) [inline test/global_uid_test.rb:41]:
```suggestion
      (1..10).each do |index|
```

**Pull request merged**